### DISCUSSION TITLE
High Availability Deployment

### DISCUSSION THREAD

**Discussion by iammattmartin** [postalserver/postal#1505] [Help with using Postal]
Would love to hear if anyone has extended out postal to a HA setup. With the postal install looking for IPs bound locally, that causes some issue.

Extending out the database/queues shouldn't be too hard but comes down to how mail is sent.

Having single reliance on a VM/server running postal is always a worry. Whilst the underlying hardware (for us) has resilience, there is still the single point of the machine instance itself (crashing badly, software corruption etc etc).

Maybe atech have some secret sauce for when they had a commercial intent for the product to span more than one running/active server.

I originally posted this in #41, so more than a re-post, wondering if anyone has managed it. It has been asked a few times so the question is to the wider community.... or maybe adamcooke (not tagging to avoid pesky emails) has any secret docs for achieving it :)

**Comment by willpower232** [comment]
It would certainly be good to get some official feedback on the matter.

Postal will send from any IP it can see by default so I presume you're trying to use IP pools so you're worried about IP addresses from that perspective?

It is probably not recommended behaviour but I have had a little mess around with the Procfile to turn off some parts of Postal and fit everything I wanted to test in a low-resources test server and I'm interested in running high availability ingress at least so our systems can at least queue up the email, even if it isn't sent immediately.

**Comment by iammattmartin** [comment]
Pretty much (in reference to the IPs). We only use internally so would like a specific brand to go out via specific IP(s). i.e some sort of IP matrix where you assign an IP each node to use for each brand. That itself is probably fairly complex.

However, for anyone just using postal as an app based mailer, that probably isn't too much of a concern. Although it would still be nice to have pools of IPs on each node to be specific to marketing, transactional and X, Y, Z mails.

**Comment by adamcooke** [comment]
I've been playing with running Postal on Kubernetes which means that potential issues with an HA deployment will be raised quite quickly. I'll let you know how I get on.

**Comment by iammattmartin** [comment]
@adamcooke, thanks for the update. Sounds exciting! :)

**Comment by mike-dube** [comment]
@adamcooke Any progress on that awesome matters?

Kubernetes is the way to go :)

**Comment by willpower232** [comment]
It should work for the most part however there are reports of problems with newer OpenSSL versions (#999 and linked issues) so you may encounter some problems but it depends which features you are using.

**Comment by willpower232** [comment]
That repository also includes YAML files which look like they would work directly with kubectl and could be modified to suit your particular configuration style.

As you can see from those files, you essentially just need to run this repo with a config directory appropriately filled in and preferably a separate deployment for each command.

**Comment by willpower232** [comment]
I believe it is capable of proxying port 25 to the SMTP server deployment but I am not sure how it would work if you want to use port 25 for different applications in the same cluster, you may need a second load balancer or whatever to give you a separate route if that makes sense.

**Comment by mfilotto** [comment]
@adamcooke A standard helm chart with official image on docker hub would be awesome

**Comment by Wouter0100** [comment]
I was looking into the Hippo repository, but all links to Hippo seems to be offline. What is Hippo and is there some documentation regarding it? Otherwise a Helm chart would be awesome.

**Reply by batonac** [reply]
Apparently: https://docs.hippofactory.dev/

**Reply by adamcooke** [reply]
It was [this](https://github.com/adamcooke/hippo) but we don't use that any longer.

**Comment by uhlhosting** [comment]
@adamcooke any news on Postsl with K8s?

**Reply by adamcooke** [reply]
Not from me I'm afraid.

**Reply by Wouter0100** [reply]
We decided not to go with Kubernetes at this time. It is fairly difficult to be able to use the IP pooling feature together with K8s. This would require significant changes in Postal and various custom routing options, including some sort of redundant NAT. 

Now we managed to make everything redundant based on Keepalived. In our case, we have assigned multiple IPv4 addresses to each pool - and those IPv4 addreses are distributed over multiple active keepalived instances to spread the load. If one of them fails, another keepalived instance will start processing those IPv4 addreses. No changes in Postal required. 

The above is only for the workers. Everything else we run just fine on K8s.

**Reply by uhlhosting** [reply]
@Wouter0100 what methods you used to port to k8s helm? So you are saying basically it will work, without the IP Pooling feature?

**Reply by uhlhosting** [reply]
It seems https://kubernetes.io/docs/concepts/cluster-administration/networking/ has been growing a lot since I last checked it. By now k8s supports in several ways IP Pools. And tbh I think of it as something to dig more, was just wanted to see how the community is around this.

**Reply by Wouter0100** [reply]
@uhlhosting without IP pooling you don't need access to the host networking, see [here](https://github.com/postalserver/install/blob/ce3ddfa66e18a7509ebec9c7aebf84fcd70675df/templates/docker-compose.yml#L22) for the Docker-compose reference and [here](https://github.com/postalserver/postal/issues/1420) for a discussion. But, I won't recommend running it without IP pooling feature in K8s - as in that case it'll send the e-mails as the K8s nodes public IPv4 or IPv6 addreses, which should be dynamic - because of the nature of K8s. 

And yes, for the others - we created a very simple helm chart. I'm looking into open sourcing it, but can't promise it. A StatefulSet for the `requeuer` and `cron` and the others are ReplicaSets (using Deployments).

**Reply by jgillich** [reply]
@willpower232 Any news on releasing your helm chart? I'd like to create one too and having a starting point would be great

**Reply by willpower232** [reply]
I have no helm chart in progress or otherwise I'm afraid

**Reply by jgillich** [reply]
Oh I'm sorry, I pinged the wrong person. That question was meant for @Wouter0100

**Reply by Wouter0100** [reply]
@jgillich unfortunately not yet.

**Reply by smolinari** [reply]
@Wouter0100 - Chalk me up as also interested in a Helm chart. 

Scott

**Reply by Pacerino** [reply]
Just pinging here to see if @Wouter0100 released his helm chart?

We're moving everything to Kubernetes.

**Reply by Wouter0100** [reply]
Unfortunately not yet. We're planning on redoing a lot at some point, after which we may be able to publish it.

**Reply by adamcooke** [reply]
We will have an official one in the not too distant future.

**Reply by Pacerino** [reply]
@Wouter0100 Thanks for your reply!
@adamcooke That sound great! I'm looking forward for it :)

**Comment by cten** [comment]
@willpower232  Does postal need any local permanent disk storage (not DB)? Trying to determine if I should even try to deploy to K8s.

**Reply by Wouter0100** [reply]
No, it does not.

**Reply by cten** [reply]
To clarify, only the "worker" sends out emails?

**Reply by Wouter0100** [reply]
Yes, we have various workers running on VM's started currently manually and do failover using keepalived. All other parts run in K8s.

See [here](https://github.com/postalserver/postal/discussions/1505#discussioncomment-1851294) for an earlier explanation.

**Reply by willpower232** [reply]
I don't think the rails error log is output to stdout but you shouldn't need persistent disk storage with the official image

**Comment by BorjaRafolsMartinez** [comment]
Anyone ever managed to get IP pools working with HA?

**Reply by masihfathi** [reply]
Did you find a way for config of haproxy?

**Reply by adamcooke** [reply]
We run one of our Postal clusters behind HAProxy without any issues. It basically works out of the box. Postal even supports "Proxy Protocol" to pass through IPs properly.

**Comment by uhlhosting** [comment]
@adamcooke how is the helm repo for postal, is it a reality?

**Comment by JumpyWizard404** [comment]
So we created a version of postal in HA with the following setup:
5 Hosts (3X MariaDB + Maxscale for automatic failover, 2X Docker Hosts for postal that both connect to mariadb via maxscale)

We put the API behind Nginx + HaProxy Loadbalancer since we are sending all of our emails via api and ip pools are shared between the two docker hosts, we are sending around 10m emails a month at the moment and the setup has worked reasonably well.
